File: 343/343_version2.java
class Solution {
    private int[] cache;
    public int integerBreak(int n) {
        cache = new int[n+1];
        cache[1] = 1;cache[2] = 1;
        return recur(n);
    }
    
    private int recur(int n){
        if(n<=2) return cache[n];
        if(cache[n] > 0) return cache[n];
        int res = 0;
        for(int temp = 1;temp<=n/2+1;temp++){
            int option1 = temp * (n-temp);
            int option2 = temp * recur(n-temp);
            res = Math.max(res,Math.max(option1,option2));
        }
        cache[n] = res;
        return cache[n];
    }
}

File: 343/343_version2.py
class Solution:
    def integerBreak(self, n: int) -> int:
        self.cache = [0 for _ in range(n+1)]
        self.cache[1] = 1
        self.cache[2] = 1
        return self.recur(n)
    
    def recur(self,n):
        if n<=2:
            return self.cache[n]
        elif self.cache[n] > 0:
            return self.cache[n]
        temp = 1
        res = 0
        while temp <= n//2+1:
            	option1 = temp * self.recur(n-temp)
		option2 = temp * (n-temp)
		res = max([res,option1,option2])
		temp+=1
        self.cache[n] = res
        return self.cache[n]
